[Safe Buffers] Serialize unsafe_buffer_usage pragmas

ziqingluo-90 opened the pull request:
A pair of `#pragma clang unsafe_buffer_usage begin/end` pragmas marks a warning-opt-out region.  The begin and end locations (opt-out regions) are stored in preprocessor instances (PP) and will be queried by the `-Wunsafe-buffer-usage` analyzer.

The commit adds serialization and de-serialization implementations for the stored regions.  Basically,  the serialized representation of the regions of a PP is a (ordered) sequence of source location encodings.   During serialization, it only serializes regions of the current translation unit.  Regions from loaded files are not serialized.  During de-serialization, regions from loaded files are stored by their source files.   When later one queries if a loaded location `l` is in an opt-out region,  PP looks up regions by the source file of `l`.  

The reported issue at upstream: https://github.com/llvm/llvm-project/issues/90501
rdar://124035402

ziqingluo-90 committed:
fix safe buffer opt-out region serialization

haoNoQ commented on clang/lib/Lex/Preprocessor.cpp:
```diff
@@ -1551,6 +1567,58 @@ bool Preprocessor::isPPInSafeBufferOptOutRegion(SourceLocation &StartLoc) {
 
+SmallVector<SourceLocation, 64>
+Preprocessor::serializeSafeBufferOptOutMap() const {
+  assert(!InSafeBufferOptOutRegion &&
```
We require the pragma to be closed before the end of TU right? And that's a hard error, not a warning?

ziqingluo-90 replied:
It's a warning.

haoNoQ replied:
Yeah in this case we need to support the situation when the warning was ignored.

ziqingluo-90 replied:
Oh,  I'm sorry,  they are not warnings, they are errors: https://github.com/llvm/llvm-project/blob/c79690040acf5bb3d857558b0878db47f7f23dc3/clang/lib/Lex/Pragma.cpp#L1265

haoNoQ commented on clang/lib/Lex/Preprocessor.cpp:
```diff
@@ -1551,6 +1567,58 @@ bool Preprocessor::isPPInSafeBufferOptOutRegion(SourceLocation &StartLoc) {
+    assert(!SourceMgr.isLocalFileID(FID) && "Expected a pch/module file");
+    // Here we assume that
+    // `SourceMgr.getFileLoc(begin) == SourceMgr.getFileLoc(end)`.
+    // Though it may not hold in very rare and strange cases, i.e., a pair of
```
If it may happen it shouldn't be an assert. If we really don't support this scenario it should probably be a compilation error, even if it only says "we don't support this yet sorry". (Can we emit compilation errors from here?)

ziqingluo-90 replied:
yes, we can emit front-end errors.  See this newly added test: https://github.com/llvm/llvm-project/pull/92031/files#diff-697faf822726557e5d62bcf106fc6cb03ff333b30b0a42ec192d93e6e12fd8fa

haoNoQ reviewed (commented):
I'm very happy that this is going somewhere! Everything makes sense to me but I also don't know a lot about this stuff.

> During serialization, it only serializes regions of the current translation unit. Regions from loaded files are not serialized.

Hmm how does this work? When a precompiled header includes another precompiled header, do they both get loaded independently and then you get to combine the maps?

ziqingluo-90 commented:
> I'm very happy that this is going somewhere! Everything makes sense to me but I also don't know a lot about this stuff.
> 
> > During serialization, it only serializes regions of the current translation unit. Regions from loaded files are not serialized.
> 
> Hmm how does this work? When a precompiled header includes another precompiled header, do they both get loaded independently and then you get to combine the maps?

For loading multiple PCHs, it is a chain.  Suppose `TU` includes pch `B` and `B` includes pch `C`.   So one would compile and emit PCHs starting from `C`, then do it for `B` with `C` being loaded.  When compile `A`, it includes pch `B` from the command line and clang actually loads `C` and `B` separately.

It is the same story for modules, except that modules form a DAG instead of a chain.   Clang will load each module file in an order with respect to the DAG.   So for each module/PCH emitted by clang,  only their own opt-out regions need to be serialized.

danakj commented:
I am curious so I will ask: why did this take an approach of serializing the output of the pragmas (the maps) instead of serializing the pragmas themselves as AST nodes, so that the maps would be constructed in the same way as from parsing the text directly? that's how other pragmas were serialized that I have seen.

haoNoQ commented:
> serializing the pragmas themselves as AST nodes

These pragmas don't really translate very well into the AST. Similarly to `#pragma clang diagnostic`, they can be placed weirdly "across" other AST nodes, like:
```
#pragma clang unsafe_buffer_usage begin
void foo() {
#pragma clang unsafe_buffer_usage end
}
```
We could support it in the same way that clang supports Duff's Device but I don't think this is really a good idea. We ultimately want them to be textual, a preprocessor-level construct, because that's what security audits are looking for: an extremely clear sign that this is where unchecked pointer operations live. So that they could tell it without looking at anything else in the code.

ziqingluo-90 committed:
fix the test command for lazily loading modules

danakj commented:
Thanks for the explanation :)

ziqingluo-90 committed:
De-serialize an unwelcomed opt-out region is an error

Throw an error if clang deserializes a safe buffer opt-out region that
begins and ends at different files.

ziqingluo-90 commented:
Just realized how stupid I am (O_o)---the current implementation only compares source locations with opt-out regions that are in the same source file, for all source locations from loaded files.   This is not going to work for cases like below.
```
// in a pch header:
#pragma clang unsafe_buffer_usage begin
#include "other_header.h"  // -Wunsafe-buffer-usage in "other_header.h" is NOT suppressed !
#pragma clang unsafe_buffer_usage end
```

Add [WIP] back to this PR.  More work is needed.

ziqingluo-90 committed:
Support multi-file safe buffer opt-out regions in PCHs

Figured out how to identify a loaded AST from a source location.  With
that, we can eliminate the restriction that a de-serialized safe
buffer opt-out region cannot span over multiple files. (Of course
those files need to be in the same TU.)

ziqingluo-90 commented:
Learned more about how loaded files & ASTs are managed in clang.   Now opt-out regions in a loaded AST can span over multiple files (within one TU).  They now behave in a consistent way when they are from loaded files or local files.

ChuanqiXu9 commented on clang/test/Modules/safe_buffers_optout.cpp:
```diff
@@ -0,0 +1,41 @@
+// RUN: %clang_cc1 -fmodules -fimplicit-module-maps -verify -fmodules-cache-path=%t -fmodule-map-file=%S/Inputs/SafeBuffers/safe_buffers_test.modulemap -I %S/Inputs/SafeBuffers %s\
+// RUN:     -x c++ -std=c++20 -Wunsafe-buffer-usage
+
+#include "test_sub1.h"
```
Now we prefer using `split-file` in tests if there are multiple inputs. You can find examples in the same folder.

ziqingluo-90 replied:
Thank you for the comment!   `split-file` is really neat!!

ChuanqiXu9 reviewed (commented):
Serialization changes look good to me. I'll leave the formal approve chance to others.

ChuanqiXu9 commented on clang/lib/Serialization/ASTWriter.cpp:
```diff
@@ -2439,6 +2440,12 @@ void ASTWriter::WritePreprocessor(const Preprocessor &PP, bool IsModule) {
 
+  // Write the safe buffer opt-out region map in PP
+  for (SourceLocation &S : PP.serializeSafeBufferOptOutMap())
+    AddSourceLocation(std::move(S), Record);
```
Generally we won't move source location.

ziqingluo-90 committed:
Use split-file for tests

haoNoQ commented on clang/lib/Lex/Preprocessor.cpp:
```diff
@@ -1483,26 +1484,40 @@ void Preprocessor::emitFinalMacroWarning(const Token &Identifier,
+      return SourceMgr.isBeforeInTranslationUnit(Map.back().first, Loc);
+    return false;
+  };
 
```
Ok @ziqingluo-90 has just walked me through this offline and I think I understand what's happening now.

What was unobvious to me: shouldn't we `TestInMap` over the entire stack of includes? And the answer is:
- Regular textual `#include`s will just show up in the current map normally; there's no need to go up the stack.
- PCHes are going to be included with the `-include-pch` flag, not directly. There's not really much of a stack going on at this point and there's no way to "put pragmas around the compiler flag".
- In case of modules, we don't care what this function returns deep inside a module, because we probably shouldn't be analyzing that code in the first place. Instead, the warning should be naturally enabled in the clang invocation that builds the module itself, and that's where the warning would be emitted, taking only the current module into account.
  - As a matter of fact, as of today we're analyzing that nested code. But we shouldn't be, that's not how modules are intended to work. We consider it a bug and have a separate plan to fix this. Right now this means that the warning will be emitted multiple times for the same line of code (both by module compilation and by .cpp file compilation, and possibly by everything in between). And the worst that could happen if this function misbehaves, is that some of these duplicates will disappear. Which is annoying but harmless.

Let's add a comment explaining this!

haoNoQ commented on clang/lib/Basic/SourceManager.cpp:
```diff
@@ -1911,6 +1911,20 @@ SourceManager::getDecomposedIncludedLoc(FileID FID) const {
+
+  auto [FID, Ignore] = getDecomposedLoc(Loc);
+  const FileID *FirstFID =
+      llvm::lower_bound(LoadedSLocEntryAllocBegin, FID, std::greater<FileID>{});
```
This is quite magical to me because I don't understand how FileIDs are assigned and in what order they show up in the `LoadedSLocEntryAllocBegin` list. But that's probably ok, this looks like completely normal `SourceManager` code to me.

haoNoQ commented:
Ok I think I'm completely happy with the patch now. It makes perfect sense to me and appears to be doing the right thing. Thank you @ziqingluo-90 for figuring this out! Let's add that comment and land.

ziqingluo-90 committed:
Make tests and comments more meaningful and hence more useful!

haoNoQ commented on clang/lib/Basic/SourceManager.cpp:
```diff
@@ -1911,18 +1911,22 @@ SourceManager::getDecomposedIncludedLoc(FileID FID) const {
          "The failure to find the first FileID of a "
          "loaded AST from a loaded source location was unexpected.");
-  return *FirstFID;
+  return FirstFID->getHashValue();
```
Hmm can we go back to returning `FileID` here? We really shouldn't rely on the fact that the _hash_ value is equal to the  actual thing we're looking for.

ziqingluo-90 committed:
Use FileID instead of its' hash value

haoNoQ reviewed (approved):
Ok LGTM!! Thanks again for figuring this all out!
